core/src/sigverifier/bls_sigverifier.rs

```diff
@@ -142,6 +227,218 @@ impl BLSSigVerifier {
+        else {
+            return false;
+        };
+
+        aggregate_bls_pubkey
+            .verify_signature(&cert_to_verify.cert_message.signature, &signed_payload)
+            .unwrap_or(false)
+    }
+
+    fn verify_base3_certificate(
+        &self,
+        cert_to_verify: &CertToVerify,
+        bit_vec1: &BitVec<u8, Lsb0>,
+        bit_vec2: &BitVec<u8, Lsb0>,
+        key_to_rank_map: &Arc<BLSPubkeyToRankMap>,
+    ) -> bool {
+        let Some((vote1, vote2)) =
+            certificate_to_vote_messages_base3(&cert_to_verify.cert_message.certificate)
+        else {
+            return false;
+        };
+
+        let Ok(signed_payload1) = bincode::serialize(&vote1) else {
+            return false;
+        };
+        let Ok(signed_payload2) = bincode::serialize(&vote2) else {
+            return false;
+        };
+
+        let messages_to_verify: Vec<&[u8]> = vec![&signed_payload1, &signed_payload2];
+
+        // 1. Aggregate the two sets of public keys separately from the two bitmaps.
+        let Some(agg_pk1) = aggregate_keys_from_bitmap(bit_vec1, key_to_rank_map) else {
+            return false;
+        };
+        let Some(agg_pk2) = aggregate_keys_from_bitmap(bit_vec2, key_to_rank_map) else {
+            return false;
+        };
+
+        let pubkeys_affine: Vec<BlsPubkey> = vec![agg_pk1.into(), agg_pk2.into()];
+        let pubkey_refs: Vec<&BlsPubkey> = pubkeys_affine.iter().collect();
+
+        SignatureProjective::par_verify_distinct_aggregated(
+            &pubkey_refs,
+            &cert_to_verify.cert_message.signature,
+            &messages_to_verify,
+        )
+        .unwrap_or(false)
+    }
+}
+
+#[derive(Debug)]
+struct VoteToVerify<'a> {
+    vote_message: VoteMessage,
+    bls_pubkey: &'a BlsPubkey,
+    packet: PacketRefMut<'a>,
+}
+
+impl VoteToVerify<'_> {
+    fn payload_slice(&self) -> &[u8] {
+        // The bincode serialization of `ConsensusMessage::Vote(VoteMessage)` has the layout:
+        // [ 4-byte Enum Discriminant | Vote Payload | 96-byte Signature | 2-byte Rank ]
+        const BINCODE_ENUM_DISCRIMINANT_SIZE: usize = 4;
+
+        // Calculate the end of the `Vote Payload` by subtracting the known size of the trailing
+        // fields (Signature and Rank) from the total packet size
+        let vote_payload_end = self
+            .packet
+            .meta()
+            .size
+            .saturating_sub(std::mem::size_of::<BlsSignature>())
+            .saturating_sub(std::mem::size_of::<u16>());
+
+        &self.packet.data(..).unwrap()[BINCODE_ENUM_DISCRIMINANT_SIZE..vote_payload_end]
+    }
+}
+
+struct CertToVerify<'a> {
+    cert_message: CertificateMessage,
+    packet: PacketRefMut<'a>,
+}
+
+fn certificate_to_vote_message_base2(certificate: &Certificate) -> Option<Vote> {
+    match certificate {
+        Certificate::Notarize(slot, hash) => Some(Vote::new_notarization_vote(*slot, *hash)),
```

Comment:
I'm thinking whether we should add a helper function near `certificate_limits_and_vote_types`, so if this changes we remember to change everything together.

I feel this whole mapping is getting complicated enough it should be an independent object, but that can happen in a later PR.